misc/sockets.rb

```diff
@@ -0,0 +1,345 @@
+# frozen_string_literal: true
+
+require 'socket'
+require 'timeout'
+
+module TestPuma
+  module SktPrepend
+    def get_body(path = nil, dly: nil, len: nil)
+      fast_write get_req(path, dly: dly, len: len)
+      read_body
+    end
+
+    def get_response(path = nil, dly: nil, len: nil)
+      fast_write get_req(path, dly: dly, len: len)
+      read_response
+    end
+
+    def read_body(time_out = 10)
+      response = read_response time_out
+      response.split("\r\n\r\n").last
+    end
+
+    def read_response(time_out = 10)
+      content_length = nil
+      response = ''.dup
+      Timeout.timeout(time_out) do
+        loop do
+          response << sysread(16_384)
+          if (content_length ||= (t = response[/Content-Length: (\d+)/i , 1]) ? t.to_i : nil)
+            return response if response.split("\r\n\r\n", 2).last.length == content_length
+          else
+            return response if response.split("\r\n\r\n", 2).length > 1
+          end
+          sleep 0.0005
+        end
+      end
+    end
+
+    def get(path = nil, dly: nil, len: nil)
+      fast_write get_req(path, dly: dly, len: len)
+    end
+
+    def write(str)
+      fast_write str
+    end
+
+    def <<(str)
+      fast_write str
+    end
+
+    def fast_write(str)
+      n = 0
+      while true
+        begin
+          n = syswrite str
+        rescue Errno::EAGAIN, Errno::EWOULDBLOCK => e
+          if !IO.select(nil, [io], nil, 5)
+            raise e
+          end
+
+          retry
+        rescue Errno::EPIPE, SystemCallError, IOError => e
+          raise e
+        end
+
+        return if n == str.bytesize
+        str = str.byteslice(n..-1)
+      end
+    end
+
+    private
+
+    def get_req(path = nil, dly: nil, len: nil)
+      req = "GET /#{path} HTTP/1.1\r\n".dup
+      req << "Dly: #{dly}\r\n" if dly
+      req << "Len: #{len}\r\n" if len
+      req << "\r\n"
+    end
+  end
+
+  class SktTCP < ::TCPSocket
+    prepend SktPrepend
+  end
+
+  if Object.const_defined? :UNIXSocket
+    class SktUnix < ::UNIXSocket
+      prepend SktPrepend
+    end
+  end
+
+  if !Object.const_defined?(:Puma) || ::Puma.ssl?
+    require 'openssl'
+    class SktSSL < ::OpenSSL::SSL::SSLSocket
+      prepend SktPrepend
+    end
+  end
+
+  module Sockets
+
+    DARWIN = !!RUBY_PLATFORM[/darwin/]
+
+    HOST = TestPuma.const_defined?(:SvrBase) ? SvrBase::HOST : '127.0.0.1'
```

Comment:
What's `SvrBase`?

Reply:
I'm working on a more DSL based framework for the tests.  `SvrBase` is a parent class for `SvrPOpen` & `SvrInProc`, which are the two classes for starting servers.  `socket.rb` and `ci_test.ru` are also part of the framework.

`socket.rb` is the client socket class, and `ci_test.ru` is a rackup that works with `socket.rb`. It allows one to set the size of the response body and a delay in the app via headers.

I'm trying to make the framework work both for CI and generic testing/benching one might want to do locally.  It's also setup to allow one to throw a lot of clients at Puma, and easily change their parameters (bind type, body size, etc).

I'm trying to get it stable (without retry) and also need to add sending multiple requests with a single client...